data_science/archive_exploration/02 - exploring trees.ipynb

```diff
@@ -0,0 +1,353 @@
+    "                      .tolist())\n",
+    "codes_as_str"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We'll now clean up the split those codes into lists of parts, using each `\\` as a delimiter."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "codes_as_list = sorted(list(map(lambda x: re.split('/|\\.', x.strip()), \n",
+    "                                codes_as_str)))\n",
+    "codes_as_list"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Lets zip these strings and lists together into a dictionary - this is going to be useful in a minute..."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "codes = dict(zip(codes_as_str, codes_as_list))\n",
+    "len(codes)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Graph theory and NetworkX\n",
+    "Seeing these lists in sorted order and thinking of them as implicit descriptions of a POSIX-style directory/folder structure, it's easy to see how the archive is divided into a hierarchy. All we need to do now is re-encode that hierarchy explicitly. To do that, we'll use a python package called NetworkX, which allows us to do a lot of basic graph theory (and some complex stuff too).  \n",
+    "As we learned in the last notebook, almost all of graph theory is based on something called an _adjacency matrix_, which describes the links between each node in the network. Each column/row represents a single node, and their intersection represents the strength of the interaction between them. Let's construct an empty matrix which we'll then populate based on the `AltRefNo` information."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "adjacency = pd.DataFrame(data=0, \n",
+    "                         index=codes_as_str, \n",
+    "                         columns=codes_as_str)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We'll now iterate through all of the possible `AltRefNo` string/list pairs. If the pair look like they have a parent/child relationship (the child's `AltRefNo` list is one element longer than the parent's and contains the full parent `AltRefNo`), then we change the element from a `0` to a `1`, drawing a connection between them."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "for str_1, list_1 in tqdm(codes.items()):\n",
+    "    for str_2, list_2 in codes.items():\n",
+    "        if ((len(list_2) == len(list_1) + 1) & \n",
+    "            (list_1 == list_2[:len(list_1)])):\n",
+    "            adjacency[str_1][str_2] = 1"
+   ]
```

Comment:
It feels like itertools and/or your Cartesian product work might be useful here.

I don’t know how fast this is in practice, but if it’s slow, maybe sort the AltRefNos? Then you can skip large parts of the matrix.

Using the child/parent terminology from the above paragraph would make this code clearer to follow. For example:

```python
for parent_id, parent_ls in tqdm(codes.items()):
    for child_id, child_ls in codes.items():
        if (
            len(child_ls) == len(parent_ls) + 1 and
            parent_ls == child_ls[:len(parent_ls)]
        ):
            adjacency[parent_id][child_id] = 1
```

Also I think you can simplify the condition:

```
if child_ls == parent_ls + [child_ls[-1]]
```

but up to you if you think that’s neater.